code/Extension/UserFormFileExtension.php

```diff
@@ -0,0 +1,72 @@
+<?php
+
+namespace SilverStripe\UserForms\Extension;
+
+use SilverStripe\Assets\File;
+use SilverStripe\Assets\Folder;
+use SilverStripe\ORM\DataExtension;
+use SilverStripe\ORM\DataObject;
+use SilverStripe\ORM\Queries\SQLUpdate;
+use SilverStripe\UserForms\Control\UserDefinedFormController;
+use SilverStripe\UserForms\Model\Submission\SubmittedFileField;
+use SilverStripe\Versioned\Versioned;
+
+class UserFormFileExtension extends DataExtension
+{
+
+    public const USER_FORM_UPLOAD_UNKNOWN = null;
+
+    public const USER_FORM_UPLOAD_FALSE = 'f';
+
+    public const USER_FORM_UPLOAD_TRUE = 't';
+
+    private static $db = [
+        'UserFormUpload' => "Enum('f, t', null)",
+    ];
+
+    /**
+     * Check if the file is associated with a userform submission
+     * Save the result in the database as a tri-state for two reasons:
+     * a) performance - prevent the need for an extra DB query
+     * b) if in the future the userform submission is deleted and the uploaded file is not (file is orphaned),
+     *    then it is still recorded that the file was originally uploaded from a userform submission
+     *
+     * @param bool $value
+     * @see File::isUserFormUpload(), UserDefinedFormController::process()
+     */
+    public function updateIsUserFormUpload(&$value): void
+    {
+        /** @var File|Versioned $file */
+        $file = $this->owner;
+        if ($file->UserFormUpload != self::USER_FORM_UPLOAD_UNKNOWN) {
+            $value = $file->UserFormUpload == self::USER_FORM_UPLOAD_TRUE;
+        } else {
+            if ($file->ClassName == Folder::class) {
+                $value = false;
+            } else {
+                $value = SubmittedFileField::get()->find('UploadedFileID', $file->ID) ? true : false;
+            }
+            $this->updateDB($value);
```

Comment:
I don't know that there's a clean way to group this work, but just noting that this could trigger up to _50_ write queries when a user loads a page in the Files UI for the first time after updating to CMS 4.6.0.

Reply:
(We could look at introducing a BuildTask to allow developers to populate this field during the upgrade if they're worried about this?)

Reply:
Yeah, the on-going requirement of "do something with data on upgrade" is a thing.

BuildTask is the defacto method, though:
- I think people will be getting BuildTask fatigue after all of the recent security issues that included BuildTask to alter their data.
- BuildTask's are also inconsistent in that only a percentage of install will every have them actually run.
- This feature is significantly lower 'value' than a task that patches potentially insecure data, so people will be less inclined to even run it

For now, I think we should just manually performance test this to see if it's an issue worth solving